native/explorer/src/datatypes.rs

```diff
@@ -350,92 +351,122 @@ fn encode_datetime_series<'b>(s: &Series, time_unit: TimeUnit, env: Env<'b>) ->
     let items: Vec<NIF_TERM> = s
         .datetime()
         .unwrap()
-        .downcast_iter()
-        .flat_map(move |iter| {
-            iter.into_iter().map(move |opt_v| {
-                opt_v
-                    .copied()
-                    .map(|x| match time_unit {
-                        TimeUnit::Milliseconds => {
-                            let naive_datetime = timestamp_to_datetime(x * 1000);
-                            encode_datetime!(
-                                naive_datetime,
-                                datetime_struct_keys,
-                                calendar_iso_c_arg,
-                                datetime_module_atom,
-                                env
-                            )
-                        }
-                        TimeUnit::Microseconds => {
-                            let naive_datetime = timestamp_to_datetime(x);
-                            encode_datetime!(
-                                naive_datetime,
-                                datetime_struct_keys,
-                                calendar_iso_c_arg,
-                                datetime_module_atom,
-                                env
-                            )
-                        }
-                        _ => unreachable!(),
-                    })
-                    .encode(env)
-                    .as_c_arg()
-            })
+        .into_iter()
+        .map(|opt_v| {
+            opt_v
+                .map(|x| match time_unit {
+                    TimeUnit::Milliseconds => {
+                        let naive_datetime = timestamp_to_datetime(x * 1000);
+                        encode_datetime!(
+                            naive_datetime,
+                            datetime_struct_keys,
+                            calendar_iso_c_arg,
+                            datetime_module_atom,
+                            env
+                        )
+                    }
+                    TimeUnit::Microseconds => {
+                        let naive_datetime = timestamp_to_datetime(x);
+                        encode_datetime!(
+                            naive_datetime,
+                            datetime_struct_keys,
+                            calendar_iso_c_arg,
+                            datetime_module_atom,
+                            env
+                        )
+                    }
+                    _ => unreachable!(),
+                })
+                .encode(env)
+                .as_c_arg()
         })
         .collect();
 
     unsafe { Term::new(env, make_list(env.as_c_arg(), &items)) }
 }
 
+fn encode_utf8_series<'b>(s: &Series, env: Env<'b>) -> Term<'b> {
+    let utf8 = s.utf8().unwrap();
+    let mut items: Vec<usize> = Vec::with_capacity(utf8.len());
```

Comment:
Looks great, I tried to make this work without pushing to an array, instead returning the map value but couldn't figure it out from my quick look.

We should also add the nil atom here so we don't have to re-encode it later.

```rust
let nil = atom::nil().to_term(env).as_c_arg()
...
                None => items.push(nil),
```